downstairs/src/extent.rs

```diff
@@ -0,0 +1,829 @@
+}
+
+/**
+ * Produce a PathBuf that refers to the completed directory we use for
+ * a given extent "number".  This directory is generated (see below) when
+ * all the files needed for repairing an extent have been copied to their
+ * final destination and everything has been flushed.
+ * The actual generation of this directory will be done by renaming the
+ * replace directory.
+ */
+pub fn completed_dir<P: AsRef<Path>>(dir: P, number: u32) -> PathBuf {
+    extent_dir(dir, number)
+        .join(extent_file_name(number, ExtentType::Data))
+        .with_extension("completed")
+}
+
+/**
+ * Remove directories associated with repair except for the replace
+ * directory. Replace is handled specifically during extent open.
+ */
+pub fn remove_copy_cleanup_dir<P: AsRef<Path>>(dir: P, eid: u32) -> Result<()> {
+    for f in [copy_dir, completed_dir] {
+        let d = f(&dir, eid);
+        if Path::new(&d).exists() {
+            std::fs::remove_dir_all(&d)?;
+        }
+    }
+    Ok(())
+}
+
+impl Extent {
+    fn get_iov_max() -> Result<usize> {
+        let i: i64 = sysconf(SysconfVar::IOV_MAX)?
+            .ok_or_else(|| anyhow!("IOV_MAX returned None!"))?;
+        Ok(i.try_into()?)
+    }
+
+    /**
+     * Open an existing extent file at the location requested.
+     * Read in the metadata from the first block of the file.
+     */
+    pub fn open(
+        dir: &Path,
+        def: &RegionDefinition,
+        number: u32,
+        read_only: bool,
+        log: &Logger,
+    ) -> Result<Extent> {
+        /*
+         * Store extent data in files within a directory hierarchy so that
+         * there are not too many files in any level of that hierarchy.
+         */
+        let path = extent_path(dir, number);
+
+        // If there was an unfinished copy, then clean it up here (??)
+        remove_copy_cleanup_dir(dir, number)?;
+
+        // If the replace directory exists for this extent, then it means
+        // a repair was interrupted before it could finish.  We will continue
+        // the repair before we open the extent.
+        let replace_dir = replace_dir(dir, number);
+        if !read_only && Path::new(&replace_dir).exists() {
+            info!(
+                log,
+                "Extent {} found replacement dir, finishing replacement",
+                number
+            );
+            move_replacement_extent(dir, number as usize, log)?;
+        }
+
+        // Pick the format for the downstairs files
+        //
+        // Right now, this only supports SQLite-flavored Downstairs
+        let inner = {
+            let mut sqlite_path = path.clone();
+            sqlite_path.set_extension("db");
+            if sqlite_path.exists() {
+                let inner = extent_inner_sqlite::SqliteInner::open(
+                    &path, def, number, read_only, log,
+                )?;
+                Box::new(inner)
+            } else {
+                panic!("no SQLite file present at {sqlite_path:?}");
```

Comment:
Good catch, restored the previous error in a6316f90cfc3ff85ae2250a6f2fb808d7d279a59

(Once we add the non-SQLite backend, this behavior will change again, because a missing `.db` file will indicate that we should load the file as a raw extent)